modepy/nodes.py

```diff
@@ -320,65 +354,121 @@ def legendre_gauss_lobatto_tensor_product_nodes(dims, n):
+def _(space: PN):
     from pytools import \
             generate_nonnegative_integer_tuples_summing_to_at_most as gnitsam
-    if shape.dims == 0:
-        return ((0,),)
-    else:
-        return tuple(gnitsam(order, shape.dims))
+    return tuple(gnitsam(space.order, space.spatial_dim))
 
 
-@get_unit_nodes.register(Simplex)
-def _(shape: Simplex, order: int):
-    import modepy as mp
-    return mp.warp_and_blend_nodes(shape.dims, order)
+@equispaced_nodes_for_space.register(PN)
+def _(space: PN, shape: Simplex):
+    if not isinstance(shape, Simplex):
+        raise NotImplementedError((type(space).__name__, type(shape).__name))
+    if space.spatial_dim != shape.dim:
+        raise ValueError("spatial dimensions of shape and space must match")
 
-# }}}
+    return (np.array(node_tuples_for_space(space), dtype=np.float64)
+            / space.order*2 - 1).T
+
+
+@edge_clustered_nodes_for_space.register(PN)
+def _(space: PN, shape: Simplex):
+    if not isinstance(shape, Simplex):
+        raise NotImplementedError((type(space).__name__, type(shape).__name))
+    if space.spatial_dim != shape.dim:
+        raise ValueError("spatial dimensions of shape and space must match")
+
+    return warp_and_blend_nodes(space.spatial_dim, space.order)
 
 
-# {{{ hypercube
+@random_nodes_for_shape.register(Simplex)
+def _(shape: Simplex, nnodes: int, rng=None):
+    if rng is None:
+        rng = np.random
 
-@get_node_count.register(Hypercube)
-def _(shape: Hypercube, order: int):
-    return (order + 1)**shape.dims
+    result = np.zeros((shape.dim, nnodes))
+    nnodes_obtained = 0
+    while True:
+        new_nodes = rng.uniform(-1.0, 1.0, size=(shape.dim, nnodes-nnodes_obtained))
+        new_nodes = new_nodes[:, new_nodes.sum(axis=0) < 2-shape.dim]
+        nnew_nodes = new_nodes.shape[1]
+        result[:, nnodes_obtained:nnodes_obtained+nnew_nodes] = new_nodes
+        nnodes_obtained += nnew_nodes
 
+        if nnodes_obtained == nnodes:
+            return result
 
-@get_node_tuples.register(Hypercube)
-def _(shape: Hypercube, order: int):
+# }}}
+
+
+# {{{ QN
+
+@node_tuples_for_space.register(QN)
+def _(space: QN):
     from pytools import \
             generate_nonnegative_integer_tuples_below as gnitb
-    if shape.dims == 0:
-        return ((0,),)
-    else:
-        return tuple(gnitb(order, shape.dims))
+    return tuple(gnitb(space.order, space.spatial_dim))
 
 
-@get_unit_nodes.register(Hypercube)
-def _(shape: Hypercube, order: int):
-    import modepy as mp
-    return mp.legendre_gauss_lobatto_tensor_product_nodes(shape.dims, order)
+@equispaced_nodes_for_space.register(QN)
+def _(space: QN, shape: Hypercube):
+    if not isinstance(shape, Hypercube):
+        raise NotImplementedError((type(space).__name__, type(shape).__name))
+    if space.spatial_dim != shape.dim:
+        raise ValueError("spatial dimensions of shape and space must match")
 
-# }}}
+    return (np.array(node_tuples_for_space(space), dtype=np.float64)
```

Comment:
Should this not be the same as `tensor_product_nodes(dim, equidistant_nodes(1, order))`?

I guess the problem is that there's now an inconsistency between the `node_tuples` and `tensor_product_nodes` order, since the nodes got flipped.

Reply:
Out of curiosity: how'd this get resolved?